File: clojure/anagram/src/anagram.clj
(ns anagram
  (:require [clojure.string :as str]))


(defn equal-frequencies?
  ([first second]
   (= (frequencies (str/lower-case first)) (frequencies (str/lower-case second)))))


(defn is-same-word [word candidate]
  (= (str/lower-case word) (str/lower-case candidate)))


(defn anagrams-for
  ([word candidates]
   (filterv #(and
                  (not (is-same-word word %1))
                  (equal-frequencies? word %1))
             candidates)))

File: ocaml/anagram/anagram.ml
open Base

let sort_string (s : string) : string =
  s
  |> String.lowercase
  |> String.to_list
  |> List.sort ~compare:Char.compare
  |> String.of_char_list

let is_anagram ( word:string ) ( candidate:string ) =
  not String.(equal (lowercase word) (lowercase candidate)) && String.(equal (sort_string word) (sort_string candidate))

let anagrams ( target: string ) ( candidates : string list ) : string list =
  List.filter candidates ~f:(fun candidate -> is_anagram target candidate)
